### PR TITLE
BE (Groups/Trends): Allow aggregating by groups

### PR DESCRIPTION
**Pull request by macobo**
## Changes

This PR adds support for aggregating by groups in trends.

Note that special care needed to be taken to account for group-less items. This
was handled by adding a special filter if aggregating by groups.

Also introduces _more_ methods to conveniently snapshot test clickhouse queries \o/

New schema change example: https://github.com/PostHog/posthog/pull/6894/files#diff-5fc2a8195db636c3cf54da6376b3b958f46fe951f469819696ef8863c53e027cR912-R927

Under every entity we now support: `math: "unique_groups"` with `math_group_type_index` field set.

## How did you test this code?

See PR test.

Some amount of bugginess is expected

### CODE REVIEW

File: ee/clickhouse/models/test/test_filters.py [hunk 1]
```diff
@@ -160,6 +161,28 @@ def test_simplify_entities(self):
...
+                    {
+                        "type": "events",
+                        "id": "$pageview",
+                        "math": "unique_group",
+                        "math_property": None,
+                        "math_group_type_index": 2,
+                        "custom_name": None,
+                        "order": None,
+                        "name": "$pageview",
+                        "properties": [{"key": "$group_2", "operator": "is_not", "value": "", "type": "event"},],
```
Comment (neilkakkar) [inline ee/clickhouse/models/test/test_filters.py:180]: Q: Groups don't allow empty string values as group keys?

File: ee/clickhouse/queries/trends/util.py [hunk 2]
```diff
@@ -22,21 +22,31 @@
...
-def process_math(entity: Entity) -> Tuple[str, str, Dict[str, str]]:
+def process_math(entity: Entity) -> Tuple[str, str, Dict[str, Any]]:
     aggregate_operation = "count(*)"
     join_condition = ""
-    params = {}
+    params: Dict[str, Any] = {}
     if entity.math == "dau":
         join_condition = EVENT_JOIN_PERSON_SQL
         aggregate_operation = "count(DISTINCT person_id)"
+    if entity.math == "unique_group":
```
Comment (neilkakkar) [inline ee/clickhouse/queries/trends/util.py:32]:
```suggestion
    elif entity.math == "unique_group":
```

File: ee/clickhouse/queries/trends/util.py [hunk 3]
```diff
@@ -22,21 +22,31 @@
...
+        if entity.math_group_type_index is None:
+            raise ValidationError(
+                {"math_group_type_index": "This field is required when `math` is set to unique_group."}, code="required"
+            )
+
+        aggregate_operation = f"count(DISTINCT JSONExtractString(properties, %(e_group_{entity.index}_math)s))"
+        params[f"e_group_{entity.index}_math"] = f"$group_{entity.math_group_type_index}"
```
Comment (neilkakkar) [inline ee/clickhouse/queries/trends/util.py:39]: small nit: extract param out to a variable so we're sure there's no string interpolation mistakes here

Comment (neilkakkar) [reply ee/clickhouse/queries/trends/util.py:39]: (and below)

File: ee/clickhouse/util.py [hunk 4]
```diff
@@ -78,3 +82,25 @@ def tearDown(self):
...
+    Snapshots are automatically saved in a __snapshot__/*.ambr file.
+    Update snapshots via --update-snapshots.
+    """
+
+    @wraps(fn)
+    def wrapped(self, *args, **kwargs):
+        with self.capture_select_queries() as queries:
+            fn(self, *args, **kwargs)
+
+        for query in queries:
```
Comment (neilkakkar) [inline ee/clickhouse/util.py:102]: (Aside: Can I get the docs to saving multiple snapshots for a single test? Didn't find it on quick glance at syrupy)

A test can have multiple queries, does this snapshotting store multiple values if necessary?

File: posthog/models/filters/mixins/simplify.py [hunk 5]
```diff
@@ -45,13 +46,16 @@ def _simplify_entity(
...
 
         EntityClass = ExclusionEntity if entity_type == "exclusions" else Entity
 
         entity = EntityClass(entity_params)
-        return EntityClass(
-            {**entity_params, "properties": self._simplify_properties(team, entity.properties, **kwargs)}
-        ).to_dict()
+        properties = self._simplify_properties(team, entity.properties, **kwargs)
+        if entity.math == "unique_group":
+            properties.append(Property(key=f"$group_{entity.math_group_type_index}", value="", operator="is_not",))
```
Comment (neilkakkar) [inline posthog/models/filters/mixins/simplify.py:56]: this is very clever!

**Review by neilkakkar — APPROVED** [review]

File: ee/clickhouse/util.py (hunk 4 above)
Comment (macobo) [reply ee/clickhouse/util.py:102]: It does - you can actually see it in some places in one of my PRs. It stats appending `.1`, `.2` etc to the snapshot name.

File: ee/clickhouse/models/test/test_filters.py (hunk 1 above)
Comment (macobo) [reply ee/clickhouse/models/test/test_filters.py:180]: Good sanity check. Yes, they don't :)

File: ee/clickhouse/queries/trends/util.py (hunk 2 above)
Comment (macobo) [reply ee/clickhouse/queries/trends/util.py:32]: Great catch :sweat_smile:

File: ee/clickhouse/util.py (hunk 4 above)
Comment (macobo) [reply ee/clickhouse/util.py:102]: Example: https://github.com/PostHog/posthog/pull/6904/files#diff-4d3a7764d124ac1edf998e86a264ea81ac45b7761f819f72e13af10b2f8591fcR80

**Pull request merged**